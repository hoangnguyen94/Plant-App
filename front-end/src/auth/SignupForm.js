import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import Alert from "../common/Alert";

/** Signup form.
 *
 * Shows form and manages update to state on changes.
 * On submission:
 * - calls signup function prop
 * - redirects to /plants route
 *
 * Routes -> SignupForm -> Alert
 * Routed as /signup
 */

const SignupForm = ( { signup } ) =>
{
    const history = useHistory();
    const [ formData, setFormData ] = useState( {
        username: "",
        password: "",
        firstName: "",
        lastName: "",
        email: "",
    } );
    const [ formErrors, setFormErrors ] = useState( [] );

    console.debug(
        "SignupForm",
        "signup=", typeof signup,
        "formData=", formData,
        "formErrors=", formErrors,
    );

    /** Handle form submit:
    *
    * Calls login func prop and, if successful, redirect to /homepage.
    */

    async function handleSubmit(evt) {
        evt.preventDefault();
        let result = await signup( formData );
        if (result.success) {
            history.push( "/homepage" );
        } else {
            setFormErrors( result.errors );
        };
        };

    /** Update form data field */
    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData( data => ( { ...data, [ name ]: value } ) );
    };

    return (
        <div className="SignupForm">
            <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
                <h2 className="mb-3">Sign Up</h2>
                <div className="card">
                    <div className="card-body">
                        <form onSubmit={handleSubmit}>
                            <div className="form-group">
                                <label htmlFor="username">Username</label>
                                <input
                                    id="username"
                                    name="username"
                                    className="form-control"
                                    value={formData.username}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="form-group">
                                <label htmlFor="password">Password</label>
                                <input
                                    id="password"
                                    type="password"
                                    name="password"
                                    className="form-control"
                                    value={formData.password}
                                    onChange={handleChange}
                                />
                            </div>

                            <div className="form-group">
                                <label htmlFor="firstName">First name</label>
                                <input
                                    id="firstName"
                                    name="firstName"
                                    className="form-control"
                                    value={formData.firstName}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="form-group">
                                <label htmlFor="lastName">Last name</label>
                                <input
                                    id="lastName"
                                    name="lastName"
                                    className="form-control"
                                    value={formData.lastName}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="form-group">
                                <label htmlFor="email">Email</label>
                                <input
                                    id="email"
                                    type="email"
                                    name="email"
                                    className="form-control"
                                    value={formData.email}
                                    onChange={handleChange}
                                />
                            </div>

                            {formErrors.length
                                ? <Alert type="danger" messages={formErrors} />
                                : null
                            }

                            <button
                                type="submit"
                                className="btn btn-primary float-right"
                                onSubmit={handleSubmit}
                            >
                                Submit
                            </button>
                        </form>
                    </div>
                </div>
            </div>
        </div>
    );
}

export default SignupForm;
