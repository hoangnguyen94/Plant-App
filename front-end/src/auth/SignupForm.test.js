import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import SignupForm from './SignupForm';

describe('SignupForm component', () => {
  test('renders the signup form component', () => {
    const { container } = render(<SignupForm signup={() => {}} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('handles form submission successfully', async () => {
    const history = createMemoryHistory();
    const signupMock = jest.fn(() => Promise.resolve({ success: true }));
    const { getByLabelText, getByText } = render(
      <Router history={history}>
        <SignupForm signup={signupMock} />
      </Router>
    );

    const usernameInput = getByLabelText('Username');
    const passwordInput = getByLabelText('Password');
    const firstNameInput = getByLabelText('First name');
    const lastNameInput = getByLabelText('Last name');
    const emailInput = getByLabelText('Email');
    const submitButton = getByText('Submit');

    fireEvent.change(usernameInput, { target: { value: 'testuser' } });
    fireEvent.change(passwordInput, { target: { value: 'testpassword' } });
    fireEvent.change(firstNameInput, { target: { value: 'Test' } });
    fireEvent.change(lastNameInput, { target: { value: 'Tester' } });
    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    fireEvent.click(submitButton);

    expect(signupMock).toHaveBeenCalledWith({
      username: 'testuser',
      password: 'testpassword',
      firstName: 'Test',
      lastName: 'Tester',
      email: 'test@example.com',
	} );
	
    expect(history.location.pathname).toBe('/');
  });

  test('handles form submission with errors', async () => {
    const history = createMemoryHistory();
    const signupMock = jest.fn(() =>
      Promise.resolve({ success: false, errors: ['Invalid username', 'Invalid password'] })
    );

    const { getByLabelText, getByText, getByRole } = render(
      <Router history={history}>
        <SignupForm signup={signupMock} />
      </Router>
    );

    const usernameInput = getByLabelText('Username');
    const passwordInput = getByLabelText('Password');
    const firstNameInput = getByLabelText('First name');
    const lastNameInput = getByLabelText('Last name');
    const emailInput = getByLabelText('Email');
    const submitButton = getByText('Submit');

    fireEvent.change(usernameInput, { target: { value: 'testuser' } });
    fireEvent.change(passwordInput, { target: { value: 'testpassword' } });
    fireEvent.change(firstNameInput, { target: { value: 'Test' } });
    fireEvent.change(lastNameInput, { target: { value: 'Tester' } });
    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    fireEvent.click(submitButton);

    expect(signupMock).toHaveBeenCalledWith({
      username: 'testuser',
      password: 'testpassword',
      firstName: 'Test',
      lastName: 'Tester',
      email: 'test@example.com',
    });
    expect(history.location.pathname).toBe('/');
  });
});
