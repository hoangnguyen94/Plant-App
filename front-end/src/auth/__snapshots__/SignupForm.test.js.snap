// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm component renders the signup form component 1`] = `
<div
  class="SignupForm"
>
  <div
    class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
  >
    <h2
      class="mb-3"
    >
      Sign Up
    </h2>
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form>
          <div
            class="form-group"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              class="form-control"
              id="username"
              name="username"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              class="form-control"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="firstName"
            >
              First name
            </label>
            <input
              class="form-control"
              id="firstName"
              name="firstName"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="lastName"
            >
              Last name
            </label>
            <input
              class="form-control"
              id="lastName"
              name="lastName"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="email"
            >
              Email
            </label>
            <input
              class="form-control"
              id="email"
              name="email"
              type="email"
              value=""
            />
          </div>
          <button
            class="btn btn-primary float-right"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
